src/witan/send/adroddiad/key_stage.clj

```diff
@@ -0,0 +1,94 @@
+  (-> {"early-years" {:order      0
+                      :name       "Early Years Foundation Stage"
+                      :label      "Early Years"
+                      :definition "From birth to 5 years old"
+                      :ncy-from   -4 ;; Note `witan.send.domain.academic-years/early-years` includes -5.
+                      :ncy-to     0}
+       "key-stage-1" {:order      1
+                      :name       "Key Stage 1"
+                      :label      "KS1"
+                      :definition "Age 5 to 7"
+                      :ncy-from   1
+                      :ncy-to     2}
+       "key-stage-2" {:order      2
+                      :name       "Key Stage 2"
+                      :label      "KS2"
+                      :definition "Age 7 to 11"
+                      :ncy-from   3
+                      :ncy-to     6}
+       "key-stage-3" {:order      3
+                      :name       "Key Stage 3"
+                      :label      "KS3"
+                      :definition "Age 11 to 14"
+                      :ncy-from   7
+                      :ncy-to     9}
+       "key-stage-4" {:order      4
+                      :name       "Key Stage 4"
+                      :label      "KS4"
+                      :definition "Age 14 to 16"
+                      :ncy-from   10
+                      :ncy-to     11}
+       "key-stage-5" {:order      5
+                      :name       "Key Stage 5"
+                      :label      "KS5"
+                      :definition "Age 16 to 19"
+                      :ncy-from   12
+                      :ncy-to     14}
+       "post-19"     {;; Note `witan.send.domain.academic-years` had this as `ncy-15+`.
+                      :order      6
+                      :name       "Post 19"
+                      :label      "Post 19"
+                      :definition "Post 19"
+                      :ncy-from   15
+                      :ncy-to     20}}
+      (update-vals (fn [m] (assoc m :ncys (into (sorted-set) (range (:ncy-from m) (inc (:ncy-to m)))))))
+      (as-> $ (into (sorted-map-by (partial compare-mapped-keys (update-vals $ :order))) $))))
+
+(def key-stages-ds
+  "Mastodon C National Curriculum Key Stage definitions as a dataset."
+  (as-> key-stages $
+    (map (fn [[k v]] (assoc v :abbreviation k)) $)
+    (tc/dataset $)
+    (tc/reorder-columns $ [:abbreviation])
+    (tc/set-dataset-name $ "key-stages")))
+
+
+
+;;; # Functions to manipulate key-stages
+(defn ncy->key-stage
+  "Given National Curriculum Year `x` and [optional] map of `key-stages`,
+   returns the MC abbreviation for the Key Stage containing it.
+  `ket-stages` must be a map with keys the key stage
+   and values maps containing a `:ncys` key whose value is a collection of the NCYs for that key stage.
+   Defaults to the namespace `key-stages` if not specified."
+  ([x] (ncy->key-stage x key-stages))
+  ([x key-stages]
+   (some (fn [[k {:keys [ncys]}]]
```

Comment:
`some `will short circuit on the first match, which I presume is what you want, but might not match with any random passed in `key-stages` map.

Reply:
I did want the algorithm to return the first key-stage containing the NCY if there were overlaps.